pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryModifierRule.java

```diff
@@ -32,94 +45,162 @@ public UnnecessaryModifierRule() {
         addRuleChainVisit(ASTAnnotationMethodDeclaration.class);
         addRuleChainVisit(ASTConstructorDeclaration.class);
     }
-    
+
+
+    private void reportUnnecessaryModifiers(Object data, Node node,
+                                            Modifier unnecessaryModifier, String explanation) {
+        reportUnnecessaryModifiers(data, node, Collections.singletonList(unnecessaryModifier), explanation);
+    }
+
+
+    private void reportUnnecessaryModifiers(Object data, Node node,
+                                            List<Modifier> unnecessaryModifiers, String explanation) {
+        if (unnecessaryModifiers.isEmpty()) {
+            return;
+        }
+        super.addViolation(data, node, new String[]{
+                formatUnnecessaryModifiers(unnecessaryModifiers),
+                getPrintableNodeKind(node),
+                getNodeName(node),
+                explanation.isEmpty() ? "" : ": " + explanation,
+        });
+    }
+
+
+    // TODO this should probably make it into a PrettyPrintingUtil or something.
+    private String getNodeName(Node node) {
+        return node instanceof ASTMethodDeclaration
+               ? ((ASTMethodDeclaration) node).getMethodName()
+               : node instanceof ASTMethodOrConstructorDeclaration
+                 // constructors are differentiated by their parameters, while we only use method name for methods
+                 ? ((ASTConstructorDeclaration) node).getQualifiedName().getOperation()
+                 : node instanceof ASTFieldDeclaration
+                   ? ((ASTFieldDeclaration) node).getVariableName()
+                   : node instanceof ASTResource
+                     ? ((ASTResource) node).getVariableDeclaratorId().getImage()
+                     : node.getImage();
```

Comment:
I'd explode this into if statements as we do in `getPrintableNodeKind`... may be longer, but will enhance readability